src/lib/core/cbus.c

```diff
@@ -177,6 +187,145 @@ cpipe_destroy(struct cpipe *pipe)
 	TRASH(pipe);
 }
 
+struct lcpipe *
+lcpipe_new(const char *consumer)
+{
+	struct lcpipe *pipe = xmalloc(sizeof(*pipe));
+
+	stailq_create(&pipe->input);
+
+	pipe->n_input = 0;
+	pipe->max_input = INT_MAX;
+
+	acquire_consumer(&pipe->endpoint, consumer);
+
+	return pipe;
+}
+
+/**
+ * Flush all staged messages into consumer output and wake up it after.
+ */
+static void
+lcpipe_flush(struct lcpipe *pipe)
+{
+	struct cbus_endpoint *endpoint = pipe->endpoint;
+	if (pipe->n_input == 0)
+		return;
+
+	/* Trigger task processing when the queue becomes non-empty. */
+	bool output_was_empty;
+
+	/*
+	 * We need to set a thread cancellation guard, because
+	 * another thread may cancel the current thread
+	 * (write() is a cancellation point in ev_async_send)
+	 * and the activation of the ev_async watcher
+	 * through ev_async_send will fail.
+	 */
+
+	int old_cancel_state;
+	tt_pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &old_cancel_state);
+
+	tt_pthread_mutex_lock(&endpoint->mutex);
+	output_was_empty = stailq_empty(&endpoint->output);
+	/** Flush input */
+	stailq_concat(&endpoint->output, &pipe->input);
+	tt_pthread_mutex_unlock(&endpoint->mutex);
+
+	pipe->n_input = 0;
+	if (output_was_empty) {
+		/* Count statistics */
+		rmean_collect(cbus.stats, CBUS_STAT_EVENTS, 1);
+
+		ev_async_send(pipe->endpoint->consumer, &pipe->endpoint->async);
+	}
+
+	tt_pthread_setcancelstate(old_cancel_state, NULL);
+}
+
+void
+lcpipe_flush_input(struct lcpipe *pipe)
+{
+	/** Flush may be called with no input. */
+	if (pipe->n_input > 0) {
+		lcpipe_flush(pipe);
+	}
+}
+
+/**
+ * Insert message into pipe input queue.
+ */
+static inline void
+lcpipe_insert_input(struct lcpipe *pipe, struct cmsg *msg)
+{
+	assert(msg->hop->pipe == NULL);
+	stailq_add_tail_entry(&pipe->input, msg, fifo);
+	pipe->n_input++;
+	if (pipe->n_input >= pipe->max_input) {
+		lcpipe_flush(pipe);
+	}
+}
+
+void
+lcpipe_push(struct lcpipe *pipe, struct cmsg *msg)
+{
+	lcpipe_insert_input(pipe, msg);
+	assert(pipe->n_input < pipe->max_input);
+}
+
+void
+lcpipe_push_now(struct lcpipe *pipe, struct cmsg *msg)
+{
+	lcpipe_insert_input(pipe, msg);
+	assert(pipe->n_input < pipe->max_input);
+	lcpipe_flush(pipe);
+}
+
+void
+lcpipe_delete(struct lcpipe *pipe)
+{
```

Comment:
This looks identical to `cpipe_destroy`.

Reply:
Yes it looks similar, but lcpipe_delete:
- calls `free` because lcpipe is always allocated to the heap
- cpipe_destroy requires that the producer to work inside a cord (lcpipe_delete can be called from any thread)
- cpipe_destroy call trigger_destroy (lcpipe doesnt support triggers)

Reply:
Yes, you need to do a few extra steps. But they are all easy to do. I've shown you diff in my other comment. Please, apply it and do the same for the other methods. When cpipe is doing something on top, you can just do it on top. The methods obviously won't be 100%. But the cpipe-specific code looks like it can be done before or after the work on its `base` member.